field/TinyLang/Field/Evaluator.hs

```diff
@@ -144,17 +148,29 @@ evalBinOp Mul x y = return . UniConst Field $ x `mul` y
 evalBinOp Div x y = UniConst Field <$> divEval x y
 evalBinOp BAt x y = asIntegerEval x >>= fmap (UniConst Bool) . flip atEval y . fromIntegral
 
+evalProgramUni
+    :: (MonadEvalError f m, Eq f, Field f, AsInteger f)
+    => Env (SomeUniConst f) -> Program f -> m (Env (SomeUniConst f))
+evalProgramUni env = evalStatementsUni env . unProgram
+
+evalStatementsUni
+    :: (MonadEvalError f m, Eq f, Field f, AsInteger f)
+    => Env (SomeUniConst f) -> Statements f -> m (Env (SomeUniConst f))
+evalStatementsUni env = foldM evalStatementUni env . unStatements
+
 evalStatementUni
     :: (MonadEvalError f m, Eq f, Field f, AsInteger f)
-    => Env (SomeUniConst f) -> Statement f -> Expr f a -> m (UniConst f a)
-evalStatementUni env (ELet (UniVar _ var) def) rest = do
+    => Env (SomeUniConst f) -> Statement f ->  m (Env (SomeUniConst f))
+evalStatementUni env (ELet (UniVar _ var) def) = do
     defR <- evalExprUni env def
-    evalExprUni (insertVar var (Some defR) env) rest
-evalStatementUni env (EAssert expr) rest = do
+    pure $ insertVar var (Some defR) env
+evalStatementUni env (EAssert expr) = do
     exprR <- evalExpr env expr
-    if exprR
-        then evalExprUni env rest
-        else throwError $ AssertionFailedEvalError expr
+    unless exprR $ throwError $ AssertionFailedEvalError expr
+    pure $ env
+evalStatementUni env (EFor (UniVar _ var) start end stmts) = foldM iteration env [start .. end] where
+    iteration env' i = evalStatementsUni env'' stmts where
+        env'' = insertVar var (Some $ fromInteger i) env'
```

Comment:
I'd probably just make it

```haskell
evalStatementUni
    :: (MonadEvalError f m, Eq f, Field f, AsInteger f)
    => Statement f ->  StateT (Env (SomeUniConst f)) m ()
```

and then you don't need to trace the state manually like you do here.